harness/determined/pytorch/_reducer.py

```diff
@@ -1,9 +1,176 @@
+import abc
+import logging
 from enum import Enum
-from typing import List, Optional
+from typing import Any, Callable, List, Type, Union
 
 import numpy as np
 
-import determined_common.check as check
+ReducerType = Union["Reducer", Type["MetricReducer"], Callable]
+
+
+class MetricReducer:
+    """
+    Efficiently aggregating validation metrics across a multi-slot distributed evaluation is done
```

Comment:
blocking: I know you copied this from the estimator Reducer. But I am not very clear of what are the outputs of the each step. The case "calculating a mean" isn't really helpful in understanding it.

A proposed doc:

```
Efficiently aggregating validation metrics across a multi-slot distributed evaluation is done in three steps:
 
    1. Accumulate metrics from each batch locally on each slot using :meth:`accumulate`.
    2. Communicate slot-locally accumulated metrics to all slots.
    3. On each slot, reduce metrics from all the slots to calculate the final metric using :meth:`cross_slot_reduce`.

   For example, you want to reduce the metrics by averaging them. The three steps are executed as follows:
    1. Accumulate the metrics by summing them up and the count of metrics.
    2. Communicate all the sums and counts to all the slots.
    3. On each slot, divide the sum by the count.
```

By the way, I know you have agreed on the API names but `aggregate` seems to be more accurate than `accumulate`.

Reply:
I'm happy to address the finer points of docstrings after the API is finalized.